File: domain/src/main/java/mixter/domain/identity/Session.java
package mixter.domain.identity;

import mixter.doc.Aggregate;
import mixter.doc.Projection;
import mixter.domain.DecisionProjectionBase;
import mixter.domain.Event;
import mixter.domain.EventPublisher;
import mixter.domain.identity.events.UserConnected;
import mixter.domain.identity.events.UserDisconnected;

import java.util.List;

@Aggregate
public class Session {

    private DecisionProjection projection;

    public Session(List<Event> history) {
        projection = new DecisionProjection(history);
    }

    public void logout(EventPublisher eventPublisher) {
        if (projection.active) {
            eventPublisher.publish(new UserDisconnected(projection.id, projection.userId));
        }
    }

    public SessionId getId() {
        return projection.id;
    }


    @Projection
    private class DecisionProjection extends DecisionProjectionBase {
        public SessionId id;
        public UserId userId;
        public boolean active = true;

        public DecisionProjection(List<Event> history) {
            register(UserConnected.class, this::apply);
            register(UserDisconnected.class, this::apply);
            history.forEach(this::apply);
        }

        public void apply(UserConnected event) {
            id = event.getSessionId();
            userId = event.getUserId();
        }

        public void apply(UserDisconnected event) {
            active = false;
        }
    }
}

File: domain/src/main/scala/mixter/domain/identity/Session.scala
package mixter.domain.identity

import mixter.domain.identity.event._
import mixter.domain.{Aggregate, EventPublisher}

object Session {
  private case class DecisionProjection(userId:UserId, sessionId: SessionId, active:Boolean){
    def apply(event:UserSessionEvent):DecisionProjection= event match{
      case UserDisconnected(_,_) => copy(active=false)
      case UserConnected(_,_,_) => this
    }
  }
  private object DecisionProjection{
    def of(userConnected: UserConnected)=DecisionProjection(userConnected.userId, userConnected.sessionId, active = true)
  }
}

case class Session(userConnected:UserConnected, history: Seq[UserSessionEvent]=Seq.empty) extends Aggregate{

  override type Id = SessionId
  override type AggregateEvent = UserSessionEvent
  override type InitialEvent = UserConnected

  private val projection={
    val seed = Session.DecisionProjection.of(userConnected)
    history.foldLeft(seed)((acc, ev)=>acc(ev))
  }

  def logout()(implicit ep:EventPublisher):Unit =
    if(projection.active){
      ep.publish(UserDisconnected(projection.sessionId, projection.userId))
    }

}
